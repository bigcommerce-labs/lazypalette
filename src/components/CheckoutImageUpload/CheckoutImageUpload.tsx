import React, { ChangeEvent, Component } from 'react';
import { Dispatch } from 'redux';
import uuid from 'uuid';

import { ThemeConfigChange } from '../../actions/theme';
import { State } from '../../reducers/reducers';
import { uploadImage } from '../../services/optimizedCheckout';
import { LoadingIndicator } from '../LoadingIndicator/LoadingIndicator';

import {
    CheckoutImageInputWrapper,
    CheckoutImagePreviewWrapper,
    CheckoutImageRemoveLink,
    CheckoutImageUploadError,
    CheckoutImageUploadWrapper,
    UploadButtonLabel
} from './styles';

export interface CheckoutImageUploadProps {
    imageURL?: string;
    inputId?: string;
    label: string;
    name: string;
    onChange(configChange: ThemeConfigChange): Dispatch<State>;
}

interface CheckoutImageUploadState {
    error?: Error;
    inputId: string;
    loading: boolean;
    pendingImageURL?: string;
    value: string;
}

class CheckoutImageUpload extends Component<CheckoutImageUploadProps, CheckoutImageUploadState> {
    readonly state: CheckoutImageUploadState = {
        inputId: this.props.inputId || uuid(),
        loading: false,
        value: '',
    };

    broadcastImageChange = (imageURL: string) => {
        this.props.onChange({ [this.props.name]: imageURL });
    };

    clearInput = (callback?: () => void) => {
        this.setState({
            loading: false,
            pendingImageURL: undefined,
            value: '',
        }, callback);
    };

    handleChangeImage = (event: ChangeEvent<HTMLInputElement>): void => {
        // Skip this handler if there are no files in the change event.
        if (!event.target.files) {
            return;
        }

        // Select the image from the file list, this is designed for a single image
        const image: File = event.target.files[0];

        // Set the internal input state.
        this.setState({
            error: undefined,
            loading: true,
            value: event.target.value,
        }, () => {
            // Send the upload to the server, once we get the response imageURL, we can broadcast a theme setting update
            uploadImage(this.props.name, image)
                .then((pendingImageURL: string) => {
                    this.setState({
                        loading: false,
                        pendingImageURL,
                        value: '',
                    }, () =>
                        this.broadcastImageChange(pendingImageURL)
                    );
                })
                .catch((error: Error) => {
                    this.setState({
                        error,
                        loading: false,
                        pendingImageURL: undefined,
                        value: '',
                    });
                });
        });
    };

    handleRemoveImage = () => {
        // TODO: we should confirm this using a modal, using window.confirm is not easily testable with jsdom.
        // See: https://github.com/jsdom/jsdom/issues/1843
        this.clearInput(() =>
            this.broadcastImageChange(''));
    };

    componentDidUpdate({ imageURL: prevImageURL }: CheckoutImageUploadProps) {
        if (this.props.imageURL !== prevImageURL) {
            this.clearInput();
        }
    }

    render() {
        const { imageURL: currentImageURL, label } = this.props;
        const { error, inputId, loading, pendingImageURL } = this.state;

        // If we have a pending image, we want to display that rather than the previous.
        const imageURLTemplate = pendingImageURL ? pendingImageURL : (currentImageURL || '');
        const imageURL = imageURLTemplate.replace('{:size}', '200x200');

        return (
            <CheckoutImageUploadWrapper>
                <CheckoutImageInputWrapper>
                    <UploadButtonLabel htmlFor={inputId}>
                        Upload image
                    </UploadButtonLabel>

                    <small>{label}</small>

                    <input
                        id={inputId}
                        onChange={this.handleChangeImage}
                        style={{display: 'none'}}
                        type={'file'}
                        value={this.state.value}
                    />
                </CheckoutImageInputWrapper>

                {error &&
                    <CheckoutImageUploadError>
                        There was an error uploading your image, please try again.
                    </CheckoutImageUploadError>
                }

                <CheckoutImagePreviewWrapper>
                    {loading
                        ? <LoadingIndicator />
                        : (imageURL &&
                            <div>
                                <div>
                                    <CheckoutImageRemoveLink onClick={this.handleRemoveImage}>
                                        Clear image
                                    </CheckoutImageRemoveLink>
                                </div>
                                <img src={imageURL} />
                            </div>
                        )
                    }
                </CheckoutImagePreviewWrapper>
            </CheckoutImageUploadWrapper>
        );
    }
}

export default CheckoutImageUpload;
