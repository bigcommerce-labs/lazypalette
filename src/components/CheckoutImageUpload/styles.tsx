import styled from 'styled-components';

export const UploadButtonLabel = styled.label`
    height: 2rem;
    border-radius: 2px;
    border: 1px solid ${({ theme }) => theme.colors.primary};
    padding: 0 0.75em;
    font-size: 0.75rem;
    cursor: pointer;
    display: inline-block;
    color: ${({ theme }) => theme.colors.primary};
    margin-right: 1rem;
`;

export const CheckoutImageUploadWrapper = styled.div`
    margin: 1rem 0;
`;

export const CheckoutImageInputWrapper = styled.div`
    margin: 1rem 0;
`;

export const CheckoutImagePreviewWrapper = styled.div``;

export const CheckoutImageUploadError = styled.p`
    color: ${({ theme }) => theme.colors.error};
`;

export const CheckoutImageRemoveLink = styled.a`
    color: ${({ theme }) => theme.colors.error};
`;
