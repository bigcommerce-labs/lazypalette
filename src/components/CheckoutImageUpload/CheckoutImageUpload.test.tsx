import Axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { mount, shallow } from 'enzyme';
import React from 'react';
import { ThemeProvider } from 'styled-components';

import CheckoutImageUpload from './CheckoutImageUpload';

describe('CheckoutImageUpload', () => {
    const axiosMock = new MockAdapter(Axios);
    const id = 'CheckoutImageUpload';

    beforeEach(() => {
        axiosMock.reset();
    });

    const theme = {
        colors: {
            errors: '#c0ffee',
            primary: '#BADA55',
        },
    };

    it('renders', () => {
        const mockHandler = jest.fn();
        const checkoutImageUpload = shallow(
            <CheckoutImageUpload
                label={'1000x400px recommended'}
                name={'optimizedCheckout-logo'}
                onChange={mockHandler}
                imageURL={''}
                inputId={id}
            />
        );

        expect(checkoutImageUpload).toMatchSnapshot();
    });

    describe('componentDidUpdate', () => {
        describe('when the component gets a new imageURL as props', () => {
            it('clears ', () => {
                const mockHandler = jest.fn();
                const checkoutImageUpload = shallow(
                    <CheckoutImageUpload
                        label={'1000x400px recommended'}
                        name={'optimizedCheckout-logo'}
                        onChange={mockHandler}
                        imageURL={''}
                        inputId={id}
                    />
                );

                checkoutImageUpload.setState({
                    loading: false,
                    pendingImageURL: 'https://placekitten.com/200/300',
                    value: 'some phony value',
                });

                expect(checkoutImageUpload.state().value).toEqual('some phony value');
                expect(checkoutImageUpload.state().loading).toEqual(false);
                expect(checkoutImageUpload.state().pendingImageURL).toEqual('https://placekitten.com/200/300');

                checkoutImageUpload.setProps({
                    imageURL: 'https://placekitten.com/200/300',
                });

                checkoutImageUpload.update();
                expect(checkoutImageUpload.state().value).toEqual('');
                expect(checkoutImageUpload.state().loading).toEqual(false);
                expect(checkoutImageUpload.state().pendingImageURL).toEqual(undefined);
            });
        });
    });

    describe('handleRemoveImage', () => {
        describe('when we have a previously set image', () => {
            it('there is a link to remove it', () => {
                const mockHandler = jest.fn();
                const wrapper = mount(
                    <ThemeProvider theme={theme}>
                        <CheckoutImageUpload
                            label={'1000x400px recommended'}
                            name={'optimizedCheckout-logo'}
                            onChange={mockHandler}
                            imageURL={''}
                            inputId={id}
                        />
                    </ThemeProvider>
                );
                const checkoutImageUpload = wrapper.find(CheckoutImageUpload).instance();
                checkoutImageUpload.setState({
                    loading: false,
                    pendingImageURL: 'https://placekitten.com/200/300',
                    value: '',
                });
                wrapper.update();
                expect(wrapper).toMatchSnapshot();
            });
        });

        describe('when the user clicks the remove button', () => {
            const event = {
                target: {
                    value: '',
                },
            };

            it('it clears the image state and broadcasts a change event', () => {
                const mockHandler = jest.fn();
                const wrapper = mount(
                    <ThemeProvider theme={theme}>
                        <CheckoutImageUpload
                            label={'1000x400px recommended'}
                            name={'optimizedCheckout-logo'}
                            onChange={mockHandler}
                            imageURL={''}
                            inputId={id}
                        />
                    </ThemeProvider>
                );
                const checkoutImageUpload = wrapper.find(CheckoutImageUpload).instance();
                checkoutImageUpload.setState({
                    loading: false,
                    pendingImageURL: 'https://placekitten.com/200/300',
                    value: '',
                });
                wrapper.update();

                wrapper.find('a').simulate('click', event);

                expect(mockHandler).toBeCalledWith({
                    setting: {
                        force_reload: true,
                        id: 'optimizedCheckout-logo',
                        type: 'optimizedCheckout-image',
                    },
                    value: '',
                });
                expect(checkoutImageUpload.state.value).toEqual('');
                expect(checkoutImageUpload.state.loading).toEqual(false);
                expect(checkoutImageUpload.state.pendingImageURL).toEqual(undefined);
                expect(checkoutImageUpload.state.error).toEqual(undefined);
                expect(wrapper).toMatchSnapshot();
            });
        });
    });

    describe('handleChangeImage', () => {
        describe('when there are no files in the event', () => {
            it('the onChange handler prop will not be called', () => {
                const mockHandler = jest.fn();
                const event = {
                    target: {
                        files: null,
                        value: '',
                    },
                };

                const checkoutImageUpload = shallow(
                    <CheckoutImageUpload
                        label={'1000x400px recommended'}
                        name={'optimizedCheckout-logo'}
                        onChange={mockHandler}
                        imageURL={''}
                        inputId={id}
                    />
                );
                const checkoutInput = checkoutImageUpload.find(`#${id}`);

                checkoutInput.simulate('change', event);
                expect(mockHandler).not.toBeCalled();
            });
        });

        describe('when the user upload files', () => {
            const event = {
                target: {
                    files: [
                        new File(
                            ['this is not a real file'],
                            'mock-file.png'
                        ),
                    ],
                    value: '',
                },
            };

            it('displays a loading indicator component when the image is being uploaded', () => {
                const mockHandler = jest.fn();
                const status = 200;
                const imageURL = 'https://placekitten.com/200/300';
                const response = { data: { imageUrl: imageURL } };

                axiosMock.onPost('/internalapi/v1/themeeditor/images')
                    .reply(status, response);

                const wrapper = mount(
                    <ThemeProvider theme={theme}>
                        <CheckoutImageUpload
                            label={'1000x400px recommended'}
                            name={'optimizedCheckout-logo'}
                            onChange={mockHandler}
                            imageURL={''}
                            inputId={id}
                        />
                    </ThemeProvider>
                );
                const checkoutImageUpload = wrapper.find(CheckoutImageUpload).instance();
                expect(checkoutImageUpload.state.loading).toEqual(false);

                const checkoutInput = wrapper.find(`#${id}`);
                checkoutInput.simulate('change', event);

                expect(checkoutImageUpload.state.loading).toEqual(true);
                expect(wrapper).toMatchSnapshot();
            });

            describe('when we get an api error', () => {
                it('displays an error message and updates internal error state', (done: () => void) => {
                    const mockHandler = jest.fn();
                    const status = 502;
                    const response = 'Internal Server Error';
                    const expectedErrorMessage = 'Request failed with status code 502';

                    axiosMock.onPost('/internalapi/v1/themeeditor/images')
                        .reply(status, response);

                    const wrapper = mount(
                        <ThemeProvider theme={theme}>
                            <CheckoutImageUpload
                                label={'1000x400px recommended'}
                                name={'optimizedCheckout-logo'}
                                onChange={mockHandler}
                                imageURL={''}
                                inputId={id}
                            />
                        </ThemeProvider>
                    );
                    const checkoutImageUpload = wrapper.find(CheckoutImageUpload).instance();
                    expect(checkoutImageUpload.state.loading).toEqual(false);
                    const checkoutInput = wrapper.find(`#${id}`);

                    checkoutInput.simulate('change', event);

                    expect(checkoutImageUpload.state.value).toEqual(event.target.value);
                    expect(checkoutImageUpload.state.loading).toEqual(true);
                    expect(checkoutImageUpload.state.error).toEqual(undefined);

                    process.nextTick(() => {
                        wrapper.update();
                        expect(mockHandler).not.toBeCalled();
                        expect(checkoutImageUpload.state.value).toEqual('');
                        expect(checkoutImageUpload.state.loading).toEqual(false);
                        expect(checkoutImageUpload.state.pendingImageURL).toEqual(undefined);
                        expect(checkoutImageUpload.state.error.message).toEqual(expectedErrorMessage);
                        expect(wrapper).toMatchSnapshot();
                        done();
                    });
                });
            });

            describe('when we get a successful response', () => {
                it('the onChange handler is called with the imageURL from the API response', () => {
                    const mockHandler = jest.fn();
                    const status = 200;
                    const imageURL = 'https://placekitten.com/200/300';
                    const response = { data: { imageUrl: imageURL } };

                    axiosMock.onPost('/internalapi/v1/themeeditor/images')
                        .reply(status, response);

                    const checkoutImageUpload = shallow(
                        <CheckoutImageUpload
                            label={'1000x400px recommended'}
                            name={'optimizedCheckout-logo'}
                            onChange={mockHandler}
                            imageURL={''}
                            inputId={id}
                        />
                    );
                    const checkoutInput = checkoutImageUpload.find(`#${id}`);
                    expect(checkoutImageUpload.state().loading).toEqual(false);
                    checkoutInput.simulate('change', event);

                    expect(checkoutImageUpload.state().value).toEqual(event.target.value);
                    expect(checkoutImageUpload.state().loading).toEqual(true);
                    expect(checkoutImageUpload.state().error).toEqual(undefined);

                    process.nextTick(() => {
                        expect(mockHandler).toBeCalledWith({
                            setting: {
                                force_reload: true,
                                id: 'optimizedCheckout-logo',
                                type: 'optimizedCheckout-image',
                            },
                            value: imageURL,
                        });
                        expect(checkoutImageUpload.state().value).toEqual('');
                        expect(checkoutImageUpload.state().loading).toEqual(false);
                        expect(checkoutImageUpload.state().pendingImageURL).toEqual(imageURL);
                        expect(checkoutImageUpload.state().error).toEqual(undefined);
                    });
                });
            });
        });
    });
});
