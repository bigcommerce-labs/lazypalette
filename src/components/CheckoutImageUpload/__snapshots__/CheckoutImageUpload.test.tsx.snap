// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutImageUpload handleChangeImage when the user upload files displays a loading indicator component when the image is being uploaded 1`] = `
<ThemeProvider
  theme={
    Object {
      "colors": Object {
        "errors": "#c0ffee",
        "primary": "#BADA55",
      },
    }
  }
>
  <CheckoutImageUpload
    imageURL=""
    inputId="CheckoutImageUpload"
    label="1000x400px recommended"
    name="optimizedCheckout-logo"
    onChange={[MockFunction]}
  >
    <styled.div>
      <div
        className="sc-bwzfXH dAsRSE"
      >
        <styled.div>
          <div
            className="sc-htpNat gMzcF"
          >
            <styled.label
              htmlFor="CheckoutImageUpload"
            >
              <label
                className="sc-bdVaJa bgvXfT"
                htmlFor="CheckoutImageUpload"
              >
                Upload image
              </label>
            </styled.label>
            <small>
              1000x400px recommended
            </small>
            <input
              id="CheckoutImageUpload"
              onChange={[Function]}
              style={
                Object {
                  "display": "none",
                }
              }
              type="file"
              value=""
            />
          </div>
        </styled.div>
        <styled.div>
          <div
            className="sc-bxivhb bUZGRr"
          >
            <Component>
              <div>
                Loading...
              </div>
            </Component>
          </div>
        </styled.div>
      </div>
    </styled.div>
  </CheckoutImageUpload>
</ThemeProvider>
`;

exports[`CheckoutImageUpload handleChangeImage when the user upload files when we get an api error displays an error message and updates internal error state 1`] = `
<ThemeProvider
  theme={
    Object {
      "colors": Object {
        "errors": "#c0ffee",
        "primary": "#BADA55",
      },
    }
  }
>
  <CheckoutImageUpload
    imageURL=""
    inputId="CheckoutImageUpload"
    label="1000x400px recommended"
    name="optimizedCheckout-logo"
    onChange={[MockFunction]}
  >
    <styled.div>
      <div
        className="sc-bwzfXH dAsRSE"
      >
        <styled.div>
          <div
            className="sc-htpNat gMzcF"
          >
            <styled.label
              htmlFor="CheckoutImageUpload"
            >
              <label
                className="sc-bdVaJa bgvXfT"
                htmlFor="CheckoutImageUpload"
              >
                Upload image
              </label>
            </styled.label>
            <small>
              1000x400px recommended
            </small>
            <input
              id="CheckoutImageUpload"
              onChange={[Function]}
              style={
                Object {
                  "display": "none",
                }
              }
              type="file"
              value=""
            />
          </div>
        </styled.div>
        <styled.p>
          <p
            className="sc-ifAKCX iAGchg"
          >
            There was an error uploading your image, please try again.
          </p>
        </styled.p>
        <styled.div>
          <div
            className="sc-bxivhb bUZGRr"
          />
        </styled.div>
      </div>
    </styled.div>
  </CheckoutImageUpload>
</ThemeProvider>
`;

exports[`CheckoutImageUpload handleRemoveImage when the user clicks the remove button it clears the image state and broadcasts a change event 1`] = `
<ThemeProvider
  theme={
    Object {
      "colors": Object {
        "errors": "#c0ffee",
        "primary": "#BADA55",
      },
    }
  }
>
  <CheckoutImageUpload
    imageURL=""
    inputId="CheckoutImageUpload"
    label="1000x400px recommended"
    name="optimizedCheckout-logo"
    onChange={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "optimizedCheckout-logo": "",
            },
          ],
        ],
      }
    }
  >
    <styled.div>
      <div
        className="sc-bwzfXH dAsRSE"
      >
        <styled.div>
          <div
            className="sc-htpNat gMzcF"
          >
            <styled.label
              htmlFor="CheckoutImageUpload"
            >
              <label
                className="sc-bdVaJa bgvXfT"
                htmlFor="CheckoutImageUpload"
              >
                Upload image
              </label>
            </styled.label>
            <small>
              1000x400px recommended
            </small>
            <input
              id="CheckoutImageUpload"
              onChange={[Function]}
              style={
                Object {
                  "display": "none",
                }
              }
              type="file"
              value=""
            />
          </div>
        </styled.div>
        <styled.div>
          <div
            className="sc-bxivhb bUZGRr"
          />
        </styled.div>
      </div>
    </styled.div>
  </CheckoutImageUpload>
</ThemeProvider>
`;

exports[`CheckoutImageUpload handleRemoveImage when we have a previously set image there is a link to remove it 1`] = `
<ThemeProvider
  theme={
    Object {
      "colors": Object {
        "errors": "#c0ffee",
        "primary": "#BADA55",
      },
    }
  }
>
  <CheckoutImageUpload
    imageURL=""
    inputId="CheckoutImageUpload"
    label="1000x400px recommended"
    name="optimizedCheckout-logo"
    onChange={[MockFunction]}
  >
    <styled.div>
      <div
        className="sc-bwzfXH dAsRSE"
      >
        <styled.div>
          <div
            className="sc-htpNat gMzcF"
          >
            <styled.label
              htmlFor="CheckoutImageUpload"
            >
              <label
                className="sc-bdVaJa bgvXfT"
                htmlFor="CheckoutImageUpload"
              >
                Upload image
              </label>
            </styled.label>
            <small>
              1000x400px recommended
            </small>
            <input
              id="CheckoutImageUpload"
              onChange={[Function]}
              style={
                Object {
                  "display": "none",
                }
              }
              type="file"
              value=""
            />
          </div>
        </styled.div>
        <styled.div>
          <div
            className="sc-bxivhb bUZGRr"
          >
            <div>
              <div>
                <styled.a
                  onClick={[Function]}
                >
                  <a
                    className="sc-EHOje bBGYrN"
                    onClick={[Function]}
                  >
                    Clear image
                  </a>
                </styled.a>
              </div>
              <img
                src="https://placekitten.com/200/300"
              />
            </div>
          </div>
        </styled.div>
      </div>
    </styled.div>
  </CheckoutImageUpload>
</ThemeProvider>
`;

exports[`CheckoutImageUpload renders 1`] = `
<styled.div>
  <styled.div>
    <styled.label
      htmlFor="CheckoutImageUpload"
    >
      Upload image
    </styled.label>
    <small>
      1000x400px recommended
    </small>
    <input
      id="CheckoutImageUpload"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
      value=""
    />
  </styled.div>
  <styled.div />
</styled.div>
`;
